BUG: automatically convert recarray dtype to np.record

ahaldane opened the pull request:
As discussed in #3581, this PR automatically converts the dtype of `np.recarray`s to `np.record`. 

To get unit tests to pass I also had to add a setattr to the `MaskedArray` class: MaskedArrays did not support assignment to dtype attribute, as demonstrated in the following examples:

```
>>> a = np.zeros(4, dtype='f4,i4')
>>> m = np.ma.array(a)
>>> m.dtype = np.dtype('f8')
>>> m   # Exception
>>> a.view(dtype='f8', type=np.ma.MaskedArray) #Exception
```

MaskedArray.setattr now catches assignments to dtype and updates the mask accordingly.

Possible issues with this PR, if anyone has any ideas: 

1) Viewing as `a.view(np.recarray)` now changes the dtype. It's not totally clear to me this won't break anything. (Example: It broke maskedarrays above, but due to a maskedaray bug).

2) Another example of the last point: Views are not reversible.

```
>>> a = np.zeros(4, 'f4,i4')
>>> b = a.view(np.recarray).view(np.ndarray)
>>> b.dtype
dtype((numpy.record, [('f0', '<f4'), ('f1', '<i4')]))
```

3) Note (not really a problem): Attempt to create record arrays of non-structured type does not set the dtype to `np.record` (since it is not possible to do so). This is not new to this PR.

```
>>> a = np.zeros(5, 'f8')
>>> np.rec.array(a)
array([ 0.,  0.,  0.,  0.], 
      dtype=float64).view(numpy.recarray)
```

Benefits:
- Goal is to improve results in #3581

I still need to finish writing unit tests

ahaldane committed:
BUG: automatically convert recarray dtype to np.record

Viewing an ndarray as a np.recarray now automatically converts
the dtype to np.record.

This commit also fixes assignment to MaskedArray's dtype attribute,
fixes the repr of recarrays with non-structured dtype, and removes
recarray.view so that viewing a recarray as a non-structured dtype
no longer converts to ndarray type.

Fixes #3581

ahaldane commented:
ugh there are still two test failures, deep in MaskedArray code.

charris commented:
My sympathies ;)

mhvk commented:
@ahaldane - I hadn't thought about the reversibility. Since that cannot possibly be guaranteed anyway, I think it is not a major problem, but perhaps it is good in the documentation to mention how one would get a regular structured array out (i.e., reverse the dtype change). Is it just `rec.view(dtype=np.dtype(np.void, rec.dtype), np.ndarray)`?

ahaldane commented:
Update, here's what's new

1) Updated docs for how to reverse the view. @mhkv That makes sense. And I think the reverse view needs to be

    rec.view(rec.dtype.fields or rec.dtype, np.ndarray)

to take into account non-structured (non-void) record arrays. IE, both of the following record arrays

    rec = np.rec.array(ones(4, dtype='f4,i4')) # structured, dtype is np.void
    rec = np.rec.array(ones(4, dtype='f8'))    # unstructured, dtype is np.float64

2) While working out the view above, I noticed that non-void record arrays can be created/represented using `rec.array` despite not having `np.record` dtype, so I modified `recarray.__repr__` accordingly.

3) I removed the `recarray.view` method. The only case it differed from ndarray.view was for non-structured dtype, in which case it returned an ndarray instead of a (non-structured) recarray. But this breaks chained views:

    >>> rec.view('f8').view('f4,i4')

The user probably wants the final result to be a record array, but the middle view currently converts to ndarray. There are already a fair number of places non-structured recarrays turn up (see above), so no reason to hide them.

4) Hopefully clarified the docs for the `dtype((base_dtype, new_dtype))` form of dtype specification.


As for the MaskedArray ssues, they pass tests now but I think there may still be something wrong. I still need to think more about it.

mhvk commented:
@ahaldane - I had a quick look through the latest PR and it seems OK (but as I don't use `recarray` much myself, someone else should definitely look as well).

ahaldane commented:
Updated with extra tweaks to the MaskedArray part. (I think the recarray part is done).

I feel better about the MaskedArray changes now. As a reminder, this needed to happen to pass unit tests related to masked record arrays. 

A problem was that the masked singleton's shape was being overwritten if I wasn't careful (another example of #5806). I think my changes are safe wrt this now. As an extra precaution I also tweaked `MaskedArray.anom` by quitting early if the result is masked, to avoid any extra manipulation of masked.

I also removed code that updates the mask's shape after the dtype is updated. I am not sure what its purpose was, and removing it doesn't affect any unit tests. Maskedarray doesn't support dtype changes that change the shape anyway. Maybe I have to think more about it, but right now it seems OK to me, and is easily changed back.

mhvk commented on numpy/core/records.py:
```diff
@@ -423,6 +423,11 @@ def __new__(subtype, shape, dtype=None, buf=None, offset=0, strides=None,
+    def __array_finalize__(self, obj):
+        if type(obj) is not type(self):
+            # invoke __setattr__
+            self.dtype = self.dtype
```
@ahaldane - In principle, at this point `self.dtype is obj.dtype` -- I think it would read better to just write `self.dtype = obj.dtype`, and expand your comment. e.g.,
```
    # invoke __setattr__, which for void dtypes will do
    # self.dtype = sb.dtype((record, obj.dtype))
    self.dtype = obj.dtype
```

ahaldane replied:
I expanded the comment, but I still have to use `self.dtype` since the condition is "self.dtype is **not** obj.dtype" ;)

mhvk commented on numpy/ma/tests/test_core.py:
```diff
@@ -1388,6 +1388,21 @@ def test_flat(self):
         assert_equal(b01.mask, array([[False, False]]))
 
+    def test_assign_dtype(self):
+        #check that the mask's dtype is updated when dtype is changed
```
This test routine doesn't seem to do what the comment promises. Should there be a few assert's that check the dtype of the mask?

ahaldane commented:
Updated with better comment + expanded tests, plus added the shape checks mentioned above back in. When adding the new tests I finally figured out why the shape manipultion was there - it raises a ValueError if the dtype change cannot apply to the mask. 

Eg, when changing from 'f4,f4' to 'f8', the mask would change from `[('f0', '?'), ('f1', '?')]` to `bool`, but it is unclear how to construct the new mask values from the old mask - should you OR the old mask? Currently MA just disallows that dtype change. 'f4,f4' to 'f4' on the other hand has no such problem.

mhvk commented on numpy/core/records.py:
```diff
@@ -423,6 +423,12 @@ def __new__(subtype, shape, dtype=None, buf=None, offset=0, strides=None,
+        if type(obj) is not type(self):
+            # invoke __setattr__, which for void dtypes will do
+            # self.dtype = sb.dtype((record, self.dtype))
+            self.dtype = self.dtype
```
@ahaldane - it should be:
```
class myarray(np.ndarray):
    def __array_finalize__(self, obj):
        print(self, obj)
        print(type(self), type(obj), type(self) is type(obj))
        if obj is not None:
            print(self.dtype, obj.dtype, self.dtype is obj.dtype)

# Now the three ways we can get to __array_finalize__
myarray((4,), buffer=np.arange(4.))
# [ 0.  1.  2.  3.] None
# <class '__main__.myarray'> <class 'NoneType'>
# -->  myarray([ 0.,  1.,  2.,  3.])
mya[:2]
# [ 0.  1.] [ 0.  1.  2.  3.]
# <class '__main__.myarray'> <class '__main__.myarray'>
# float64 float64 True
# --> myarray([ 0.,  1.])
np.arange(4.).view(myarray)
# [ 0.  1.  2.  3.] [ 0.  1.  2.  3.]
# <class '__main__.myarray'> <class 'numpy.ndarray'> False
# float64 float64 True
# --> Out[15]: myarray([ 0.,  1.,  2.,  3.])
```

I think this is just a problem with your `if` statement; it should be
```
if obj is not None and type(self) is not type(obj):
    self.dtype = obj.dtype
```

ahaldane replied:
Maybe I'm being too clever, but I think we want the dtype to be set to `np.record` in the case of an explicit constructor. I still think the clause I have is the right one.

    class TestClass(np.ndarray):
        def __array_finalize__(self, obj):
            print(obj is not None and type(self) is not type(obj), 
                  type(obj) is not type(self))

    a = np.arange(10)

    t1 = a.view(TestClass)  # view fron ndarray
    t2 = TestClass(4)       # explicit constructor
    t3 = t1[:2]             # slice
    t4 = t1.view(TestClass) # view from same class


This prints out (left is your if, right is my if)

    (True, True)
    (False, True)
    (False, False)
    (False, False)

I think the right is what I want. Is that correct?

mhvk replied:
OK, if you need it to be set also for explicit construction then, yes, your case is the right one. Since this was not obvious to me, and therefore may not be obvious to a future person looking at the code, maybe write the comment as:
```
# Both when constructing a new instance (obj=None) and when viewing
# from another class -- obj.view(type(self)) -- we invoke __setattr__,
# which for void dtypes will do self.dtype = sb.dtype((record, self.dtype)).
```

mhvk commented:
@ahaldane - nice catch on the mask change with incompatible dtype. While `or`ing the mask seems OK for the example you give, it does seem safest to just disallow it, at least for now!

ahaldane commented:
@mhvk  the detailed comment is a good idea. Done.

mhvk commented:
Great, looks all OK to me!

ahaldane commented:
Don't merge, I found a problem:

    >>> r = np.rec.array(np.ones(4, dtype='f4,f4'))
    >>> r.view('f8').dtype
    dtype('float64')

will fix soon.

ahaldane commented:
Oh never mind, that is a scalar recarray. It is working as expected, I just confused myself.

But it did make me notice, perhaps a better if clause in array_finalize (as we were discussing above) is

    if self.dtype.type is not record

I think that's clearer, right?

charris commented:
Agree.

mhvk commented:
Yes, better, now the test is much more self-explanatory.

charris commented:
@ahaldane @mhvk Thanks, merged. I hope it is correct ;)

ahaldane commented:
Thanks, and thanks @mhkv for helpful comments & feedback.
